src/Middleware/CORS/src/Infrastructure/CorsMiddleware.cs

```diff
@@ -183,14 +182,30 @@ private async Task InvokeCore(HttpContext context, ICorsPolicyProvider corsPolic
             if (corsPolicy == null)
             {
                 // Resolve policy by name if the local policy is not being used
-                corsPolicy = await corsPolicyProvider.GetPolicyAsync(context, policyName);
+                var policyTask = corsPolicyProvider.GetPolicyAsync(context, policyName);
+                if (!policyTask.IsCompletedSuccessfully)
+                {
+                    return InvokeCoreAwaited(context, policyTask);
+                }
+
+                corsPolicy = policyTask.GetAwaiter().GetResult();
             }
 
+            return EvaluateAndApplyPolicy(context, corsPolicy);
+        }
+
+        private async Task InvokeCoreAwaited(HttpContext context, Task<CorsPolicy> policyTask)
```

Comment:
nit: static local method?

Reply:
Can't be static, because it references a non-static method, but we can make it a local function